core/base/mergeTreePrincipalGeodesics/MergeTreePrincipalGeodesicsBase.cpp

```diff
@@ -0,0 +1,334 @@
+    vec = std::vector<double>(size);
+    for(unsigned int i = 0; i < size; ++i)
+      vec[i] = pVec[i];
+  }
+
+  void MergeTreePrincipalGeodesicsBase::pointersToVectors(
+    std::vector<double *> &pVec,
+    std::vector<size_t> sizes,
+    std::vector<std::vector<double>> &vec) {
+    vec = std::vector<std::vector<double>>(pVec.size());
+    for(unsigned int i = 0; i < pVec.size(); ++i)
+      pointerToVector(pVec[i], sizes[i], vec[i]);
+  }
+
+  void MergeTreePrincipalGeodesicsBase::pointersToVectors(
+    std::vector<double *> &pVec,
+    size_t size,
+    std::vector<std::vector<double>> &vec) {
+    std::vector<size_t> sizes(pVec.size(), size);
+    pointersToVectors(pVec, sizes, vec);
+  }
+
+  // ----------------------------------------------------------------------------
+  // Matrix Utils
+  // ----------------------------------------------------------------------------
+  void MergeTreePrincipalGeodesicsBase::matrixDot(
+    std::vector<std::vector<double>> &m1,
+    std::vector<std::vector<double>> &m2,
+    std::vector<std::vector<double>> &newM) {
+    newM = std::vector<std::vector<double>>(
+      m1.size(), std::vector<double>(m2[0].size(), 0.0));
+    for(unsigned int i = 0; i < newM.size(); ++i)
+      for(unsigned int j = 0; j < newM[i].size(); ++j)
+        for(unsigned int k = 0; k < m1[i].size(); ++k)
+          newM[i][j] += m1[i][k] * m2[k][j];
+  }
+
+  void MergeTreePrincipalGeodesicsBase::subMatrix(
+    std::vector<std::vector<double>> &m1,
+    std::vector<std::vector<double>> &m2,
+    std::vector<std::vector<double>> &newM) {
+    newM = std::vector<std::vector<double>>(
+      m1.size(), std::vector<double>(m1[0].size()));
+    for(unsigned int i = 0; i < m1.size(); ++i)
+      for(unsigned int j = 0; j < m1[0].size(); ++j)
+        newM[i][j] = m1[i][j] - m2[i][j];
+  }
+
+  void MergeTreePrincipalGeodesicsBase::sumMatrix(
+    std::vector<std::vector<double>> &m1,
+    std::vector<std::vector<double>> &m2,
+    std::vector<std::vector<double>> &newM) {
+    newM = std::vector<std::vector<double>>(
+      m1.size(), std::vector<double>(m1[0].size()));
+    for(unsigned int i = 0; i < m1.size(); ++i)
+      for(unsigned int j = 0; j < m1[0].size(); ++j)
+        newM[i][j] = m1[i][j] + m2[i][j];
+  }
+
+  void MergeTreePrincipalGeodesicsBase::multMatrix(
+    std::vector<std::vector<double>> &m1,
+    double mult,
+    std::vector<std::vector<double>> &newM) {
+    newM = m1;
+    for(unsigned int i = 0; i < newM.size(); ++i)
+      for(unsigned int j = 0; j < newM[i].size(); ++j)
+        newM[i][j] *= mult;
+  }
+
+  void MergeTreePrincipalGeodesicsBase::transpose(
+    std::vector<std::vector<double>> &m,
+    std::vector<std::vector<double>> &newM) {
+    std::vector<std::vector<double>> mTemp(
+      m[0].size(), std::vector<double>(m.size()));
+    for(unsigned int i = 0; i < m.size(); ++i)
+      for(unsigned int j = 0; j < m[0].size(); ++j)
+        mTemp[j][i] = m[i][j];
+    newM = mTemp;
```

Comment:
It was in case we want to use the same matrix as input and output (to do the transpose inplace). We can keep this like that, or remove this feature or we can also change the loop in order to swap m[i][j] and m[j][i] at the same time and avoid copying the whole matrix at the end while still allowing to do the transpose inplace.

Reply:
The new definition now avoid copy. It is left to the used of the function.